### PR TITLE
Implement sample weights for faraday

### PR DESCRIPTION
**Pull request by shengy90**
Calculates weighted losses by:

1. For MMD loss, we have to expand the tensors based on the sample weights as MSE weights operate on a distribution level
2. For MSE loss and quantile loss, we:
    - Calculate sample-wise losses
    - Instead of using the standard reduction method (mean), we calculate the weighted mean
    - Do this by multiplying sample-wise losses by sample weights, then taking the sum of losses and divide by the sum of weights
 
Edge cases and logic are tested in `TestFaradayLosses`.

### CODE REVIEW

File: src/opensynth/models/faraday/losses.py [hunk 1]
```diff
@@ -59,12 +82,44 @@ def quantile_loss(
...
-        torch.max(
-            quantile * (y_real - y_pred), (1 - quantile) * (y_pred - y_real)
-        )
+
+    quantile_loss = torch.max(
+        quantile * (y_real - y_pred), (1 - quantile) * (y_pred - y_real)
     )
 
+    if sample_weights is not None:
+        quantile_loss = quantile_loss * sample_weights
```
Comment (Ianlmgoddard) [inline src/opensynth/models/faraday/losses.py:91]: I feel very pedantic writing this, but I think it avoids a copy of `quantile_loss`. Will leave it up to you if you want to change it. 
```suggestion
        quantile_loss *= sample_weights
```

**Comment by Ianlmgoddard** [comment]
Looks fine to me! as long as we haven't exported `MMDLoss` or use it outside of the `FaradayModel` then it should be fine!

**Review by Ianlmgoddard — APPROVED** [review]

File: src/opensynth/models/faraday/losses.py [hunk 2]
```diff
@@ -98,21 +154,23 @@ def calculate_training_loss(
...
     quantile_median_loss = (
-        quantile_loss(x_hat, x, 0.5) * quantile_median_weight
+        quantile_loss(x_hat, x, 0.5, sample_weights) * quantile_median_weight
     )
     quantile_losses = (
         quantile_upper_loss + quantile_lower_loss + quantile_median_loss
     )
 
-    total_loss = mmd_loss + mse_loss + quantile_losses
+    total_loss = mmd_loss_ + mse_loss_ + quantile_losses
```
[the remark this answers is not in the document]
Comment (shengy90) [reply src/opensynth/models/faraday/losses.py:174]: No quantile loss can never be negative because of the torch.max() operation.

File: src/opensynth/models/faraday/losses.py (hunk 1 above)
Comment (shengy90) [reply src/opensynth/models/faraday/losses.py:91]: Haha i'm the opposite, I always do x = x + y and avoid *= because of accidentally overriding other assigned objects in place. E.g.

`*= operator`
```python
a = torch.tensor([1,2,3])
b = a
a *= 2 

--> a is now torch.tensor([2,4,6])
--> but b is also now torch.tensor([2,4,6]) because b=a doesn't result in a new copy of `b` variable but share the same memory
```

`x = x * 2 explicitly`
```python
a = torch.tensor([1,2,3])
b = a
a = a * 2

--> a is torch.tensor([2,4,6])
---> b however remains torch.tensor([1,2,3])
```

I think in this case, everything is scoped within a function and there isn't any other shared objects, so it's safe, but got bitten by this weird behaviour before and have always stayed away from `*=` haha.

File: src/opensynth/models/faraday/losses.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (charlotte-avery) [reply src/opensynth/models/faraday/losses.py:174]: I think non-negative quantile looks relies on y_pred < y_real at least once in the tensor.

[the remark this answers is not in the document]
Comment (charlotte-avery) [reply src/opensynth/models/faraday/losses.py:174]: Can MMD loss also be negative?
